File: android/src/main/java/com/netrareactnative/dto/RequestOptionsDTO.kt
package com.example.netra_flutter.dto

import com.netra.library.NetraRequestConfig
import com.netra.library.NetraResponse
import com.netra.library.enums.OfflinePolicyAction

data class RequestOptionsDTO(
    val id: String,
    val url: String,
    val offlinePolicyAction: OfflinePolicyActionDTO? = null,
    val slowNetworkPolicyAction: SlowNetworkPolicyActionDTO? = null,
    val cacheOptions: CacheOptionsDTO? = null,
    val headers: Map<String, String>?,
    val cancelOnDispose: Boolean? = false,
    val body: RequestBodyDTO?= null,
) {
    companion object {
        fun fromDataModel(config: NetraRequestConfig): RequestOptionsDTO {
            return RequestOptionsDTO(
                id = config.id,
                url = config.url,
                offlinePolicyAction = config.offlinePolicy?.let {
                    OfflinePolicyActionDTO.fromDataModel(it)
                },
                slowNetworkPolicyAction = config.slowNetworkPolicy?.let {
                    SlowNetworkPolicyActionDTO.fromDataModel(it)
                },
                cacheOptions = config.cache?.let {
                    CacheOptionsDTO.fromDataModel(it)
                },
                headers = config.headers,
                cancelOnDispose = config.cancelOnDispose,
                body = config.body?.let {
                    RequestBodyDTO.fromDataModel(it)
                },
            )
        }
    }
}

File: src/internal/dto/RequestOptionsDTO.ts
import { BaseDTO } from './BaseDTO';
import { OfflinePolicyActionDTO } from './OfflinePolicyActionDTO';
import { SlowNetworkPolicyActionDTO } from './SlowNetworkPolicyActionDTO';
import { CacheOptionsDTO } from './CacheOptionsDTO';
import { RequestBodyDTO } from './RequestBodyDTO';
import { RequestOptions } from '../../models';

export class RequestOptionsDTO extends BaseDTO {
  id: string;
  url: string;
  offlinePolicyAction?: OfflinePolicyActionDTO;
  slowNetworkPolicyAction?: SlowNetworkPolicyActionDTO;
  cacheOptions?: CacheOptionsDTO;
  headers?: Map<string, string>;
  cancelOnDispose?: boolean;
  body?: RequestBodyDTO;

  constructor(
    id: string,
    url: string,
    offlinePolicyAction?: OfflinePolicyActionDTO,
    slowNetworkPolicyAction?: SlowNetworkPolicyActionDTO,
    cacheOptions?: CacheOptionsDTO,
    headers?: Map<string, string>,
    cancelOnDispose?: boolean,
    body?: RequestBodyDTO
  ) {
    super();
    this.id = id;
    this.url = url;
    this.offlinePolicyAction = offlinePolicyAction;
    this.slowNetworkPolicyAction = slowNetworkPolicyAction;
    this.cacheOptions = cacheOptions;
    this.headers = headers;
    this.cancelOnDispose = cancelOnDispose;
    this.body = body;
  }

  static fromDataModel(model: RequestOptions) {
    return new RequestOptionsDTO(
      model.id,
      model.url,
      model.offlinePolicyAction !== undefined
        ? OfflinePolicyActionDTO.fromDataModel(model.offlinePolicyAction)
        : undefined,
      model.slowNetworkPolicyAction !== undefined
        ? SlowNetworkPolicyActionDTO.fromDataModel(
            model.slowNetworkPolicyAction
          )
        : undefined,
      model.cacheOptions !== undefined
        ? CacheOptionsDTO.fromDataModel(model.cacheOptions)
        : undefined,
      model.headers,
      model.cancelOnDispose,
      model.body !== undefined
        ? RequestBodyDTO.fromDataModel(model.body)
        : undefined
    );
  }

  static fromJSON(json: string) {
    const parsedJSON = JSON.parse(json);
    return new RequestOptionsDTO(
      parsedJSON.id,
      parsedJSON.url,
      parsedJSON.offlinePolicyAction !== undefined
        ? OfflinePolicyActionDTO.fromJSON(
            JSON.stringify(parsedJSON.offlinePolicyAction)
          )
        : undefined,
      parsedJSON.slowNetworkPolicyAction !== undefined
        ? SlowNetworkPolicyActionDTO.fromJSON(
            JSON.stringify(parsedJSON.slowNetworkPolicyAction)
          )
        : undefined,
      parsedJSON.cacheOptions !== undefined
        ? CacheOptionsDTO.fromJSON(JSON.stringify(parsedJSON.cacheOptions))
        : undefined,
      parsedJSON.headers,
      parsedJSON.cancelOnDispose,
      parsedJSON.body !== undefined
        ? RequestBodyDTO.fromJSON(JSON.stringify(parsedJSON.body))
        : undefined
    );
  }

  toDataModel(): RequestOptions {
    return new RequestOptions({
      url: this.url,
      offlinePolicyAction: this.offlinePolicyAction?.toDataModel(),
      slowNetworkPolicyAction: this.slowNetworkPolicyAction?.toDataModel(),
      cacheOptions: this.cacheOptions?.toDataModel(),
      headers: this.headers,
      cancelOnDispose: this.cancelOnDispose,
      body: this.body?.toDataModel(),
    });
  }

  toJSON() {
    return {
      id: this.id,
      url: this.url,
      offlinePolicyAction: this.offlinePolicyAction?.toJSON(),
      slowNetworkPolicyAction: this.slowNetworkPolicyAction?.toJSON(),
      cacheOptions: this.cacheOptions?.toJSON(),
      headers: this.headers,
      cancelOnDispose: this.cancelOnDispose,
      body: this.body?.toJSON(),
    };
  }
}
